### ISSUE TITLE
Boolean operation which flips my fill value

### ISSUE THREAD

**Issue by jwdebelius** [pydata/sparse#297]
Hi!

This is an awesome package and I'm having fun using it to its full potential. Hopefully this is the right place for this. I'm having an issue where, after a series of boolean operations, my fill value becomes True and my data value becomes False. This limits my ability to do further calculations.

I have a sparse, boolean `n` x `n` x `m` array. 

```python
>>> import sparse as sp
>>> m = sp.COO(
...    coords=np.array([[0, 0, 0], [0, 0, 1],
...                     [0, 1, 0],
...                     [1, 0, 0],
...                     [1, 1, 0], [1, 1, 1],
...                     [2, 2, 1], 
...                     [3, 3, 0], [3, 3, 1]
...                     ]).T,
...    data=1,
...    shape=(4, 4, 2),
... )
>>> m
<COO: shape=(4, 4, 2), dtype=int64, nnz=9, fill_value=0>
```

I want to identify rows/columns where there is not any value that is present in all three coordinates by checking for the number of positions that have something present and then that this is less than the number of regions.

```python
>>> pres = (m.sum(axis=1) > 0).sum(axis=1) < 2
>>> pres
<COO: shape=(4,), dtype=bool, nnz=3, fill_value=True>
>>> pres.todense()
array([False, False,  True, False])
```

If I try to do something like flip my fill value, then I just get a matrix of zeros

```python
>>> pres.fill_value = 0
>>> pres
<COO: shape=(4,), dtype=bool, nnz=3, fill_value=0>
>>> pres.todense()
array([False, False,  False, False])
```

But, the fact that my array density flips seems like  a possible bug, or at least a parameter I'd like more control over.

**Comment by hameerabbasi** [comment]
Your second code snippet illustrates the issue.

> ```python
> >>> pres = (m.sum(axis=1) > 0).sum(axis=1) < 2
> >>> pres
> <COO: shape=(4,), dtype=bool, nnz=3, fill_value=True>
> >>> pres.todense()
> array([False, False,  True, False])
> ```

If there's a row of all-zeros, it'll sum to zero, so the resulting fill-value is zero. When you compare an array with a fill-value of zero to 2, it's going to be less most of the time.

This is the reason that we have a fill-value of `True` in the resulting array. However, you make a point that one should be able to manipulate the fill-value in the manner you show.

**Comment by jwdebelius** [comment]
Im not sure I understand why this instance in particular flips the fill value. Is this based on a density assumption under the hood that i'm missing, or is it just that some boolean operations will do the flip and others won't?
Is there a way to predict when the fill value will be inverted? 

In my case, it matters for subsequent matrix operations, for instance, I cant do `sp.dot` of the pres matrix.

**Comment by hameerabbasi** [comment]
> It would still be really nice to designate the state that should occupy the fill value without having to pass through a dense array.

Agreed, an API would be nice for this, would you mind opening an issue for that and closing this one?

**Comment by hameerabbasi** [comment]
I thought more about this, and while possible for booleans, it's impossible to change the fill-value in general.

**Comment by hameerabbasi** [comment]
Whoops, having thought a bit more, it's possible.
